Sources/PenguinStructures/ArrayStorage.swift

```diff
@@ -0,0 +1,184 @@
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//******************************************************************************
+// This file exposes reference-counted buffer types similar to the storage for
+// `Array`, but designed to be (potentially) handled through type-erased APIs
+// that are statically `Element`-type-agnostic.
+//******************************************************************************
+
+/// Effectively, the stored properties of every `ArrayStorage` instance.
+fileprivate struct ArrayHeader {
+  var count: Int
+  var capacity: Int
+}
+
+/// Contiguous storage of homogeneous elements of statically unknown type.
+///
+/// This class provides the element-type-agnostic API for ArrayStorage<T>.
+open class AnyArrayStorage {
+  /// An instance that provides the implementation of element-type-agnostic APIs
+  /// defined in this class.
+  open var implementation: AnyArrayStorageImplementation {
+    fatalError("implement me!")
+  }
+  
+  /// Appends the instance of the concrete element type whose address is `p`,
+  /// returning the index of the appended element, or `nil` if there was
+  /// insufficient capacity remaining
+  public final func appendValue(at p: UnsafeRawPointer) -> Int? {
+    implementation.appendValue_(at: p)
+  }
+
+  /// Invokes `body` with the memory occupied by initialized elements.
+  public final func withUnsafeMutableRawBufferPointer<R>(
+    _ body: (inout UnsafeMutableRawBufferPointer)->R
+  ) -> R {
+    implementation.withUnsafeMutableRawBufferPointer_(body)
+  }
+
+  /// Deinitializes all stored data.
+  deinit { implementation.deinitialize() }
+}
+
+/// Contiguous storage of homogeneous elements of statically unknown type.
+///
+/// Conformances to this protocol provide the implementations for
+/// `AnyArrayStorage` APIs.
+public protocol AnyArrayStorageImplementation: AnyArrayStorage {
+  /// Appends the instance of the concrete element type whose address is `p`,
+  /// returning the index of the appended element, or `nil` if there was
+  /// insufficient capacity remaining
+  func appendValue_(at p: UnsafeRawPointer) -> Int?
+  
+  /// Invokes `body` with the memory occupied by initialized elements.
+  func withUnsafeMutableRawBufferPointer_<R>(
+    _ body: (inout UnsafeMutableRawBufferPointer)->R
+  ) -> R
+
+  /// Deinitialize stored data
+  func deinitialize()
+}
+
+/// Contiguous storage of homogeneous elements of statically known type.
+///
+/// This protocol's extensions provide APIs that depend on the element type, and
+/// the implementations for `AnyArrayStorage` APIs.
+public protocol ArrayStorageImplementation: AnyArrayStorageImplementation {
```

Comment:
Nit: in Swift, I associate `Array`s with value semantics, but I didn't see CoW machinery here... would it be better to call this `BufferStorageImplementation` or maybe `BufferImplementation`?

Reply:
“Array” has the connotations I was looking for: contiguity and homogeneity.  “Buffer” doesn't necessarily do that.  Also, this follows the convention set in the standard library: “Storage” is a class that provides the storage (https://github.com/apple/swift/blob/master/stdlib/public/core/ContiguousArrayBuffer.swift#L71), “ContiguousArrayBuffer” is a struct that adds resizability, and “Array” wraps that to provide value semantics.

Happy to use some other terms if you're convinced they're better though.

Reply:
I'm fine with following the precedent for now, and we can think harder about this over time. LGTM!